From a9c4510dd6ab537d8de2a09766f0ebf6f481cd2e Mon Sep 17 00:00:00 2001
From: Olivier Fourdan <ofourdan@redhat.com>
Date: Tue, 4 Jun 2024 17:02:57 +0200
Subject: [PATCH 1/2] xwayland: Do not use manual redirect windows as surface
 window

While walking the window tree looking for the surface window to use, we
should ignore windows using manual redirection.

Those are to be considered as "transparent", they do not obscure
siblings or parent windows.

Signed-off-by: Olivier Fourdan <ofourdan@redhat.com>
Fixes: 3a0fc268 - xwayland: Add xwl_window::surface_window
Closes: https://gitlab.freedesktop.org/xorg/xserver/-/issues/1677
Closes: https://gitlab.freedesktop.org/xorg/xserver/-/issues/1679
---
 hw/xwayland/xwayland-window.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/hw/xwayland/xwayland-window.c b/hw/xwayland/xwayland-window.c
index 05f45e16f6..0dd1eff0a6 100644
--- a/hw/xwayland/xwayland-window.c
+++ b/hw/xwayland/xwayland-window.c
@@ -1365,6 +1365,9 @@ xwl_window_update_surface_window(struct xwl_window *xwl_window)
         if (window->drawable.depth == 32)
             continue;
 
+        if (window->redirectDraw == RedirectDrawManual)
+            continue;
+
         surface_window = window;
     }
 
-- 
GitLab


From 1a3db0a1580668a4d16c7030c6b7a854e555d89c Mon Sep 17 00:00:00 2001
From: Olivier Fourdan <ofourdan@redhat.com>
Date: Tue, 4 Jun 2024 17:26:29 +0200
Subject: [PATCH 2/2] xwayland: Stop on first unmapped child

If a child window of the same size is unmapped, we should stop walking
the tree looking for the surface window to use.

Whatever lies beneath is not visible anyway.

This also fixes an issue with the Damage list becoming corrupted when
destroying a window, because the first thing that DeleteWindow() does
is to unmap the window and crush the window tree underneath it.

Signed-off-by: Olivier Fourdan <ofourdan@redhat.com>
Fixes: 3a0fc268 - xwayland: Add xwl_window::surface_window
Closes: https://gitlab.freedesktop.org/xorg/xserver/-/issues/1680
---
 hw/xwayland/xwayland-window.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/hw/xwayland/xwayland-window.c b/hw/xwayland/xwayland-window.c
index 0dd1eff0a6..fdf27b32f5 100644
--- a/hw/xwayland/xwayland-window.c
+++ b/hw/xwayland/xwayland-window.c
@@ -1351,6 +1351,9 @@ xwl_window_update_surface_window(struct xwl_window *xwl_window)
         if (!RegionEqual(&window->winSize, &surface_window->winSize))
             break;
 
+        if (!window->mapped)
+            break;
+
         /* The surface window must be top-level for its window pixmap */
         window_pixmap = screen->GetWindowPixmap(window);
         if (window_pixmap == surface_pixmap)
-- 
GitLab

